## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#02b059c8-292e-4279-9e8c-b00fe1c938b9) |
| mint-large-bucket.sh | :x: [more...](#6a3b452b-636b-4e19-850c-36c19204cef5) |
| mint-compression-fs.sh | :x: [more...](#d152506c-d913-480a-ba9c-249c9239ffb1) |
<a name="6a3b452b-636b-4e19-850c-36c19204cef5"></a>
### 7915-61de8c5/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.61:31861
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 94ab53cf22e3:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 2 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 53 seconds
(4/14) Running aws-sdk-ruby tests ... done in 13 seconds
(5/14) Running awscli tests ... done in 2 minutes and 20 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... FAILED in 2 minutes and 16 seconds
{
  "name": "mc",
  "duration": "91640",
  "function": "test_mirror_list_objects_storage_class",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-29784 --quiet --no-color mirror --storage-class REDUCED_REDUNDANCY /mint/data myminio/mc-test-bucket-4529  >>> \n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-4529/datafile-1-MB`\n`/mint/data/datafile-0-b` -> `myminio/mc-test-bucket-4529/datafile-0-b`\n`/mint/data/datafile-1-kB` -> `myminio/mc-test-bucket-4529/datafile-1-kB`\n`/mint/data/datafile-1-b` -> `myminio/mc-test-bucket-4529/datafile-1-b`\n`/mint/data/datafile-10-MB` -> `myminio/mc-test-bucket-4529/datafile-10-MB`\n`/mint/data/datafile-1.03-MB` -> `myminio/mc-test-bucket-4529/datafile-1.03-MB`\n`/mint/data/datafile-100-kB` -> `myminio/mc-test-bucket-4529/datafile-100-kB`\n`/mint/data/datafile-10-kB` -> `myminio/mc-test-bucket-4529/datafile-10-kB`\n`/mint/data/datafile-11-MB` -> `myminio/mc-test-bucket-4529/datafile-11-MB`\n`/mint/data/datafile-129-MB` -> `myminio/mc-test-bucket-4529/datafile-129-MB`\n`/mint/data/datafile-33-kB` -> `myminio/mc-test-bucket-4529/datafile-33-kB`\n`/mint/data/datafile-5-MB` -> `myminio/mc-test-bucket-4529/datafile-5-MB`\n`/mint/data/datafile-6-MB` -> `myminio/mc-test-bucket-4529/datafile-6-MB`\n`/mint/data/datafile-5243880-b` -> `myminio/mc-test-bucket-4529/datafile-5243880-b`\n`/mint/data/datafile-65-MB` -> `myminio/mc-test-bucket-4529/datafile-65-MB`\nmc: <ERROR> Failed to copy `/mint/data/datafile-65-MB`. Please reduce your request\nmc: <ERROR> Failed to copy `/mint/data/datafile-10-MB`. Please reduce your request\nmc: <ERROR> Failed to copy `/mint/data/datafile-129-MB`. Please reduce your request\nmc: <ERROR> Failed to copy `/mint/data/datafile-11-MB`. Please reduce your request\nTotal: 233.17 MiB, Transferred: 491.99 MiB, Speed: 5.37 MiB/s"
}

Executed 6 out of 14 tests successfully.
```
<a name="d152506c-d913-480a-ba9c-249c9239ffb1"></a>
### 7915-61de8c5/mint-compression-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.61:31898
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 68707cea0cd8:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 41 seconds
(4/14) Running aws-sdk-ruby tests ... done in 2 seconds
(5/14) Running awscli tests ... done in 58 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 17 seconds
(8/14) Running minio-dotnet tests ... done in 23 seconds
(9/14) Running minio-go tests ... done in 41 seconds
(10/14) Running minio-java tests ... done in 46 seconds
(11/14) Running minio-js tests ... done in 40 seconds
(12/14) Running minio-py tests ... FAILED in 2 minutes and 10 seconds
{
  "name": "minio-py:test_presigned_get_object_expiry_5sec",
  "args": {
    "response.status-2": 200,
    "response.headers": "{\"_container\": {\"accept-ranges\": [\"Accept-Ranges\", \"bytes\"], \"content-length\": [\"Content-Length\", \"1024\"], \"content-security-policy\": [\"Content-Security-Policy\", \"block-all-mixed-content\"], \"content-type\": [\"Content-Type\", \"application/octet-stream\"], \"etag\": [\"ETag\", \"\\\"866119c1236723a2075990241d72fa5b-1\\\"\"], \"last-modified\": [\"Last-Modified\", \"Thu, 11 Jul 2019 23:24:47 GMT\"], \"server\": [\"Server\", \"MinIO/DEVELOPMENT.2019-07-11T23-00-22Z\"], \"vary\": [\"Vary\", \"Origin\"], \"x-amz-request-id\": [\"X-Amz-Request-Id\", \"15B07DE3C9224BEE\"], \"x-xss-protection\": [\"X-Xss-Protection\", \"1; mode=block\"], \"date\": [\"Date\", \"Thu, 11 Jul 2019 23:24:47 GMT\"]}}",
    "object_name": "1c9320ce-fb83-4047-8704-dcf2ab8cff42",
    "response.reason": "OK",
    "presigned_get_object_url": "http://72.28.97.61:31898/minio-py-test-ad60008e-5ec8-4d71-a42c-ec5f47ce71c8/1c9320ce-fb83-4047-8704-dcf2ab8cff42?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minio%2F20190711%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190711T232448Z&X-Amz-Expires=5&X-Amz-SignedHeaders=host&X-Amz-Signature=3ee47f3609ecbc5982dbdf34eab10488b34d0a559278fec56c96a96b626a0fdb",
    "response.status": 200,
    "response.reason-2": "OK",
    "response._body": "rsugtxmwxjvffrheiogassoeyyuxojagnyeebyjyhshlurdnnzjnmhqlomggwmzoiegvhuezoqrgfvltyxtegzevuozdgprwxgoohrytypdadfngelgbkcacoefiyphyuecxemqzkclifqcawenyhgxrihehxekpiljuclunlulhzlyxupaeyixgacdglbozjokwdzzcbuejnyprnxyyhaqutxgrcrhmysmtlnxoeuzrtefbbljnjbqkotnffkdukzwusakbuagrdowrwvpunpikwnxlcwukyinazkeskdtshizysyibsigolqrmyitzwssbhxbcnjodrsmoourevelwbgqkufedruytbfzmclpxvzfgyohudfbjporhklhpvctmunqtzzapahhfzrjnbigcxpguhmvroqaknhhjtyehpgqznsfekqpxyiwhoavedyqxnzcnlgckinlbxyuvaugrbqncmhskxghujuebpzzcadyadowwooejxaktrpxompjzrdiuyqvlnmesbnjsgvohwgmcstxttxihhxcqmshygcfzlyywfcdfywsnkltipogzofaxohermeazflcepiiqqwunwetzgphqxmqzblaspmtubkxshougcayfzhunlvzdnvuskdocgsrqwxjaqppparellsebrrwakzqxgnfqwrsuxjhekucqpbgbzkpebzshjekruclngjnbjsumwjoxxfapmqvdbcdtvibsekkhnatdtkphkhuloccrirmzkuuzekcablkzieqmbvyivezpocihtfpgenckyuxtmuqulaqvognulsbpicxnqecxojbeerupsnlldjrazeideneydfzrlubfcllnajplkpxevwjtszecqgmyfyhcxntbmamdwnbqjhhvzttuioxpwkjlbhvpatnbtbzhgrhfgvhwoptzazemfocjilsaqcjgwdoyniqttbkthgbyggexrsuhroucyrmceewgoixugljerewpddqsygfderludtid",
    "bucket_name": "minio-py-test-ad60008e-5ec8-4d71-a42c-ec5f47ce71c8",
    "response.headers-2": "{\"_container\": {\"accept-ranges\": [\"Accept-Ranges\", \"bytes\"], \"content-length\": [\"Content-Length\", \"1024\"], \"content-security-policy\": [\"Content-Security-Policy\", \"block-all-mixed-content\"], \"content-type\": [\"Content-Type\", \"application/octet-stream\"], \"etag\": [\"ETag\", \"\\\"866119c1236723a2075990241d72fa5b-1\\\"\"], \"last-modified\": [\"Last-Modified\", \"Thu, 11 Jul 2019 23:24:47 GMT\"], \"server\": [\"Server\", \"MinIO/DEVELOPMENT.2019-07-11T23-00-22Z\"], \"vary\": [\"Vary\", \"Origin\"], \"x-amz-request-id\": [\"X-Amz-Request-Id\", \"15B07DE4F3AE180A\"], \"x-xss-protection\": [\"X-Xss-Protection\", \"1; mode=block\"], \"date\": [\"Date\", \"Thu, 11 Jul 2019 23:24:52 GMT\"]}}",
    "response._body-2": "rsugtxmwxjvffrheiogassoeyyuxojagnyeebyjyhshlurdnnzjnmhqlomggwmzoiegvhuezoqrgfvltyxtegzevuozdgprwxgoohrytypdadfngelgbkcacoefiyphyuecxemqzkclifqcawenyhgxrihehxekpiljuclunlulhzlyxupaeyixgacdglbozjokwdzzcbuejnyprnxyyhaqutxgrcrhmysmtlnxoeuzrtefbbljnjbqkotnffkdukzwusakbuagrdowrwvpunpikwnxlcwukyinazkeskdtshizysyibsigolqrmyitzwssbhxbcnjodrsmoourevelwbgqkufedruytbfzmclpxvzfgyohudfbjporhklhpvctmunqtzzapahhfzrjnbigcxpguhmvroqaknhhjtyehpgqznsfekqpxyiwhoavedyqxnzcnlgckinlbxyuvaugrbqncmhskxghujuebpzzcadyadowwooejxaktrpxompjzrdiuyqvlnmesbnjsgvohwgmcstxttxihhxcqmshygcfzlyywfcdfywsnkltipogzofaxohermeazflcepiiqqwunwetzgphqxmqzblaspmtubkxshougcayfzhunlvzdnvuskdocgsrqwxjaqppparellsebrrwakzqxgnfqwrsuxjhekucqpbgbzkpebzshjekruclngjnbjsumwjoxxfapmqvdbcdtvibsekkhnatdtkphkhuloccrirmzkuuzekcablkzieqmbvyivezpocihtfpgenckyuxtmuqulaqvognulsbpicxnqecxojbeerupsnlldjrazeideneydfzrlubfcllnajplkpxevwjtszecqgmyfyhcxntbmamdwnbqjhhvzttuioxpwkjlbhvpatnbtbzhgrhfgvhwoptzazemfocjilsaqcjgwdoyniqttbkthgbyggexrsuhroucyrmceewgoixugljerewpddqsygfderludtid"
  },
  "duration": 5052,
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1242, in test_presigned_get_object_expiry_5sec\n    raise ValueError('Presigned get url failed to expire!')\nValueError: Presigned get url failed to expire!\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1913, in main\n    test_presigned_get_object_expiry_5sec(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1244, in test_presigned_get_object_expiry_5sec\n    raise Exception(err)\nException: Presigned get url failed to expire!\n",
  "function": "presigned_get_object(bucket_name, object_name, expires, response_headers, request_date)",
  "status": "FAIL",
  "message": "Presigned get url failed to expire!"
}

Executed 11 out of 14 tests successfully.
```
<a name="02b059c8-292e-4279-9e8c-b00fe1c938b9"></a>
### 7915-61de8c5/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.52:30682
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 50f51a1c70b3:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 6 minutes and 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-wrloovn2jfmamn",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 360015,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 15B07DDBF7524010, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```

--- a/cmd/http-tracer.go
+++ b/cmd/http-tracer.go
@@ -43,10 +43,16 @@ type recordRequest struct {
 	logBody bool
 	// Internal recording buffer
 	buf bytes.Buffer
+	// request headers
+	headers http.Header
+	// total bytes read including header size
+	bytesRead int
 }
 
 func (r *recordRequest) Read(p []byte) (n int, err error) {
 	n, err = r.Reader.Read(p)
+	r.bytesRead += n
+
 	if r.logBody {
 		r.buf.Write(p[:n])
 	}
@@ -54,6 +60,13 @@ func (r *recordRequest) Read(p []byte) (n int, err error) {
 		return n, err
 	}
 	return n, err
+}
+func (r *recordRequest) Size() int {
+	sz := r.bytesRead
+	for k, v := range r.headers {
+		sz += len(k) + len(v)
+	}
+	return sz
 }
 
 // Return the bytes that were recorded.
@@ -80,13 +93,17 @@ type recordResponseWriter struct {
 	statusCode int
 	// Indicate if headers are written in the log
 	headersLogged bool
+	// number of bytes written
+	bytesWritten int
 }
 
 // Write the headers into the given buffer
-func writeHeaders(w io.Writer, statusCode int, headers http.Header) {
-	fmt.Fprintf(w, "%d %s\n", statusCode, http.StatusText(statusCode))
+func (r *recordResponseWriter) writeHeaders(w io.Writer, statusCode int, headers http.Header) {
+	n, _ := fmt.Fprintf(w, "%d %s\n", statusCode, http.StatusText(statusCode))
+	r.bytesWritten += n
 	for k, v := range headers {
-		fmt.Fprintf(w, "%s: %s\n", k, v[0])
+		n, _ := fmt.Fprintf(w, "%s: %s\n", k, v[0])
+		r.bytesWritten += n
 	}
 }
 
@@ -94,7 +111,7 @@ func writeHeaders(w io.Writer, statusCode int, headers http.Header) {
 func (r *recordResponseWriter) WriteHeader(i int) {
 	r.statusCode = i
 	if !r.headersLogged {
-		writeHeaders(&r.headers, i, r.ResponseWriter.Header())
+		r.writeHeaders(&r.headers, i, r.ResponseWriter.Header())
 		r.headersLogged = true
 	}
 	r.ResponseWriter.WriteHeader(i)
@@ -102,10 +119,11 @@ func (r *recordResponseWriter) WriteHeader(i int) {
 
 func (r *recordResponseWriter) Write(p []byte) (n int, err error) {
 	n, err = r.ResponseWriter.Write(p)
+	r.bytesWritten += n
 	if !r.headersLogged {
 		// We assume the response code to be '200 OK' when WriteHeader() is not called,
 		// that way following Golang HTTP response behavior.
-		writeHeaders(&r.headers, http.StatusOK, r.ResponseWriter.Header())
+		r.writeHeaders(&r.headers, http.StatusOK, r.ResponseWriter.Header())
 		r.headersLogged = true
 	}
 	if (r.statusCode != http.StatusOK && r.statusCode != http.StatusPartialContent && r.statusCode != 0) || r.logBody {
@@ -113,6 +131,10 @@ func (r *recordResponseWriter) Write(p []byte) (n int, err error) {
 		r.body.Write(p)
 	}
 	return n, err
+}
+
+func (r *recordResponseWriter) Size() int {
+	return r.bytesWritten
 }
 
 // Calls the underlying Flush.
@@ -151,9 +173,17 @@ func getOpName(name string) (op string) {
 func Trace(f http.HandlerFunc, logBody bool, w http.ResponseWriter, r *http.Request) trace.Info {
 	name := getOpName(runtime.FuncForPC(reflect.ValueOf(f).Pointer()).Name())
 
+	// Setup a http request body recorder
+	reqHeaders := cloneHeader(r.Header)
+	reqHeaders.Set("Content-Length", strconv.Itoa(int(r.ContentLength)))
+	reqHeaders.Set("Host", r.Host)
+	for _, enc := range r.TransferEncoding {
+		reqHeaders.Add("Transfer-Encoding", enc)
+	}
+
 	var reqBodyRecorder *recordRequest
 	t := trace.Info{FuncName: name}
-	reqBodyRecorder = &recordRequest{Reader: r.Body, logBody: logBody}
+	reqBodyRecorder = &recordRequest{Reader: r.Body, logBody: logBody, headers: reqHeaders}
 	r.Body = ioutil.NopCloser(reqBodyRecorder)
 	t.NodeName = r.Host
 	if globalIsDistXL {
@@ -162,14 +192,6 @@ func Trace(f http.HandlerFunc, logBody bool, w http.ResponseWriter, r *http.Requ
 	// strip port from the host address
 	if host, _, err := net.SplitHostPort(t.NodeName); err == nil {
 		t.NodeName = host
-	}
-
-	// Setup a http request body recorder
-	reqHeaders := cloneHeader(r.Header)
-	reqHeaders.Set("Content-Length", strconv.Itoa(int(r.ContentLength)))
-	reqHeaders.Set("Host", r.Host)
-	for _, enc := range r.TransferEncoding {
-		reqHeaders.Add("Transfer-Encoding", enc)
 	}
 
 	rq := trace.RequestInfo{
@@ -199,5 +221,7 @@ func Trace(f http.HandlerFunc, logBody bool, w http.ResponseWriter, r *http.Requ
 
 	t.ReqInfo = rq
 	t.RespInfo = rs
+
+	t.CallStats = trace.CallStats{Latency: rs.Time.Sub(rq.Time), InputBytes: reqBodyRecorder.Size(), OutputBytes: respBodyRecorder.Size()}
 	return t
 }

--- a/pkg/trace/trace.go
+++ b/pkg/trace/trace.go
@@ -24,10 +24,18 @@ import (
 // Info - represents a trace record, additionally
 // also reports errors if any while listening on trace.
 type Info struct {
-	NodeName string       `json:"nodename"`
-	FuncName string       `json:"funcname"`
-	ReqInfo  RequestInfo  `json:"request"`
-	RespInfo ResponseInfo `json:"response"`
+	NodeName  string       `json:"nodename"`
+	FuncName  string       `json:"funcname"`
+	ReqInfo   RequestInfo  `json:"request"`
+	RespInfo  ResponseInfo `json:"response"`
+	CallStats CallStats    `json:"stats"`
+}
+
+// CallStats records request stats
+type CallStats struct {
+	InputBytes  int           `json:"inputbytes"`
+	OutputBytes int           `json:"outputbytes"`
+	Latency     time.Duration `json:"latency"`
 }
 
 // RequestInfo represents trace of http request

